runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/stableinput/BufferedElement.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.runners.flink.translation.wrappers.streaming.stableinput;
```

Comment:
Much of the following code isn't Flink specific? Would it be worthwhile to move that into a runner shared package and make the Flink specific pieces abstract?

Reply:
I would say much of the code is rather Flink specific. I'm not sure whether this deserves to be moved to a common runner package as other Runners might have a different approach to solving this problem.

Reply:
More or less what you have built here is a checkpointing aware write-ahead log. All the encoding for buffered elements and timers are generic, and most of `BufferingDoFnRunner` could possibly be if it could work based on `StateInternals`.